src/peep/peep.c

```diff
@@ -7772,13 +7772,385 @@ static void peep_on_enter_or_exit_ride(rct_peep *peep, int rideIndex, int flags)
+		RCT2_GLOBAL(0x00F1AEE3, uint8) = 1; // then this becomes true
+		has_voucher = true;
+	}
+
+///loc_69AF47:
+	// Check to see if the peep already has the item that they're considering buying
+	int shopItem2 = shopItem; // A separate item that's used as output
+	// The root if block you see right below is common in this function.
+	// It seems to be used to branch based on whether this shop serves either a
+	// standard or an extra item. It can probably be replaced with a bool and shopItem
+	// used directly.
+	if (shopItem >= 32) {
+		shopItem2 -= 32;
+		if (peep->item_extra_flags & (1 << shopItem2)) {
+			peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_ALREADY_GOT, shopItem);
+			return 0;
+		}
+	}
+	else {
+		if (peep->item_standard_flags & (1 << shopItem2)) {
+			peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_ALREADY_GOT, shopItem);
+			return 0;
+		}
+	}
+
+///loc_69AF6B:
+	// Checks to see if the peep is already eating or is too nauseous
+	// TODO: check to make sure all of the conditions are covered
+	if (shop_item_is_food_or_drink(shopItem)) {
+		if (peep_has_food(peep)) {
+			int food = bitscanforward(peep->item_standard_flags & STANDARD_FOOD_MASK);
+			if (food != -1)
+				peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_HAVENT_FINISHED, food);
+			else
+				peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_HAVENT_FINISHED, (bitscanforward(peep->item_extra_flags & EXTRA_FOOD_MASK) + 32));
+			return 0;
+		}
+		else {
+			if (peep->nausea >= 145)
+				return 0;
+		}
+	}
+
+///loc_69AFD3:
+	// Don't buy certain items if it's raining
+	if ((shopItem == SHOP_ITEM_BALLOON) || (shopItem == SHOP_ITEM_ICE_CREAM)
+		|| (shopItem == SHOP_ITEM_COTTON_CANDY) || (shopItem == SHOP_ITEM_SUNGLASSES)) {
+		if (RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_RAIN_LEVEL, sint8) != 0)
+			return 0;
+	}
+
+///loc_69B007:
+	// Don't buy certain items if it's too cold out
+	if ((shopItem == SHOP_ITEM_SUNGLASSES) || (shopItem == SHOP_ITEM_ICE_CREAM)) {
+		if (RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_TEMPERATURE, sint8) < 12)
+			return 0;
+	}
+
+///loc_69B03B:
+	// Possible original bug: chocolate is not considered a food item for this check.
+	if (shop_item_is_food(shopItem) && (peep->hunger < 75)) {
+		peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_NOT_HUNGRY, 0xFF); // Null arg?
+		return 0;
+	}
+
+///loc_69B069:
+	if (shop_item_is_drink(shopItem) && (peep->thirst < 75)) {
+		peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_NOT_THIRSTY, 0xFF);
+		return 0;
+	}
+
+///loc_69B097:
+	//0x14001B = standard: balloon, toy, photo, umbrella, hat, t-shirt (but not map)
+	//0x87 = photo 2/3/4, sunglasses
+	if ((shopItem == SHOP_ITEM_UMBRELLA) && (RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_RAIN_LEVEL, sint8) != 0))
+			goto loc_69B119;
+
+///loc_69B0A4:
+	if ((shopItem == SHOP_ITEM_MAP) && (peep->guest_heading_to_ride_id == 0xFF))
+			goto loc_69B119;
+
+///loc_69B0B1:
+	if ((!((shopItem >= 32 ? 0x87 : 0x14001B) & (1 << shopItem2))) && !has_voucher) {
```

Comment:
The logic between loc_69B097 and loc 69B119 is really hairy. I'm trying to maneuver my way out of using a goto, but I'm having a rough time doing it.